runtime/lua/vim/_editor.lua

```diff
@@ -715,6 +715,58 @@ function vim._on_key(buf, typed_buf)
   end
 end
 
+--- Convert UTF-32 or UTF-16 {index} to byte index.
+--- {opts} can be a table containing two properties:i
+--- * an "encoding" property of "utf-8", "utf-16" or "utf-32"
+--- * an "error" property of true or false, if error is false
+--- then then an out of range index will return byte length
+--- instead of throwing an error.
+---
+--- {opts} can also be supplied as a boolean, true for utf-16 and false for utf-32.
+--- If {opts} is not supplied, it defaults to false (UTF-32). Returns the byte index.
+---
+--- Invalid UTF-8 and NUL is treated like in |vim.str_utfindex()|.
+--- An {index} in the middle of a UTF-16 sequence is rounded upwards to
+--- the end of that sequence.
+---@param s string
+---@param index integer
+---@param opts? boolean|nil| { encoding: "utf-8"|"utf-16"|"utf-32", error?: boolean }
+---@return integer
+function vim.str_byteindex(s, index, opts)
+  vim.validate({ s = { s, 'string' }, index = { index, 'number' } })
+  local len = #s
+  if index == 0 then
+    return 0
+  end
+
+  if index > len then
+    if opts.error then
+      error('index out of range')
+    end
+    return len
+  end
+
+  opts = opts or { encoding = 'utf-32', error = true }
+  if type(opts) == 'boolean' then
+    opts = opts and { encoding = 'utf-16', error = true } or { encoding = 'utf-32', error = true }
+  end
+  local encoding = opts.encoding or 'utf-32'
+  local valid_encodings = { ['utf-8'] = true, ['utf-16'] = true, ['utf-32'] = true }
+  if not valid_encodings[encoding] then
+    error('Invalid encoding: ' .. encoding)
+  end
```

Comment:
This function should be fast, however a lot of the code here will add unnecessary overhead for typical call arguments.

Can some of this logic be bypassed if opts isn't provided?

Reply:
added an early return for a boolean or nil opts